### DISCUSSION TITLE
Middleware runs twice on server side

### DISCUSSION THREAD

**Discussion by dicedatalore** [nuxt/nuxt#24569] [Questions]
Nuxt middleware runs twice on server on startup and occasionally on refresh. This is causing problems trying to refresh my auth token on the server side. it runs twice, the second time causing an error. How can I avoid or work around this behaviour?

reproduction https://stackblitz.com/edit/github-a3tmfz
you can see in the logs when it starts up that it is run twice.

**Comment by GhostvOne** [comment]
Same here my middleware on server side only is executing twice

**Comment by Perdolique** [comment]
I'm testing global middleware that should redirect users from page `/test` to page `/`. So in my mind, the middleware should be called twice. One time before navigation to `/test` and the second time before navigation to `/` after redirecting. But it called three times.

```
redirect middleware called from /test to /test
redirect middleware called from /test to /
redirect middleware called from / to /
```

I've tried to use `external: true` from [this issue](https://github.com/nuxt/nuxt/issues/20696#issuecomment-2159965021). It helped, and now the second log string disappeared, only the first and the third strings were in logs.

I don't understand the logic here. No "external" URLs were used.

[This is an example.](https://stackblitz.com/edit/nuxt-middleware-double-calling?file=middleware%2Fredirect.global.ts) Try to navigate to `/test` manually.

Nuxt v3.12.3
